### ISSUE TITLE
[tool] Proposal to support dart define config from a json file 

### ISSUE THREAD

**Issue by hai046** [flutter/flutter#107810]
## Use case

- If `--dart-define` has many definitions, the startup command can be very long and error-prone
- It is especially difficult to maintain if there are multiple packages to switch
- Also I want these definitions to be available directly in android and iOS

We counted, our project package includes countries: Brazil, Singapore, Europe, we have a total of 36 configuration items in each region

In my app, different regions and languages use different package names for internationalization. Different package names google-ad, google-service.json, facebook, and firebase have different configurations. It is very cumbersome to switch between definitions. Centralized management and configuration are required to avoid business trips
 
So I hope ：
- Use project variables directly in android build.gradle
- Use environment variables directly in iOS

## Proposal

I have a [PR](https://github.com/flutter/flutter/pull/107089) ，hope you can give some advice, thanks ! (Because I was not familiar with the process in the early stage, I raised PR first, sorry again)

Add a new flag to support config constants pool by json file

1、Define  json file 

``` shell
flutter build apk  --define-config-json-file=cn.com.test.pkg.json
```
You can use `flutter run --help` get usage

``` shell
    --dart-define=<foo=bar>                       Additional key-value pairs that will be available as constants from the String.fromEnvironment, bool.fromEnvironment, int.fromEnvironment, and double.fromEnvironment constructors.
                                                  Multiple defines can be passed by repeating "--dart-define" multiple times.
    --define-config-json-file                     The path of a json format file where flutter define a global constant pool. Use `const String.fromEnvironment("DEFINE_CONFIG_JSON_RAW_VALUE")` can get all config json file raw
                                                  value. Json entry will be available as constants from the String.fromEnvironment, bool.fromEnvironment, int.fromEnvironment, and double.fromEnvironment constructors;the key is json
                                                  filed key,get value is json value.

```
 
for example: `cn.com.test.pkg.json`  content
``` json
{
  "facebookAppId": "xxx",
  "versionName": "1.0.0",
  "label": "xxx xxx",
  "versionCode": 1,
  "android_applicationId": "com.xxx.br",
  "googleAndroidApiKey": "xxxx",
  "googleAndroidAppId": "1:xxx:android:xxx",
  "googleAndroidMessagingSenderId": "xxx",
  "schemeHost": "first.xxx.com",
  "scheme": "xxxxx",
  "privacyProtocol": "https://xxxx.com/privacyProtocol.html",
  "teamsOfService": "https://xxx.com/teamsOfService.html",
  "androidChannel": "1OSH1PP64X",
  "iosChannel": "1J6C22IH91",
  "iosUploadCSymS": "false",
  "universalLinkDomain": "br.xxx.com"
}
```

2、Use environment in flutter

 You can get all json raw content:
``` dart
var configRaw =
        const String.fromEnvironment("DEFINE_CONFIG_JSON_RAW_VALUE", defaultValue: "{}");
var  configRaw = utf8.decode(base64Decode(configRaw));
```
Or ,you can get field by 

``` dart
var facebookAppId =
        const String.fromEnvironment("facebookAppId", defaultValue: "");
```

3、Android project 

file: `build.gradle`

``` gradle
// This then uses the project variable
    buildTypes {
        release {
            manifestPlaceholders = [
                    faceBook_appId       : facebookAppId
            ]

            resValue "string", "facebook_app_id", facebookAppId

…………
// switch profile
task switchToGoogleServices(type: Copy) {
    description = "Switches google-services config,from google-services.${android_applicationId}.json to google-services.json"
    println(description)
    from "../../precompileConfigs/google-services"
    include "google-services.${android_applicationId}.json"
    into "./"
    rename "google-services.${android_applicationId}.json", 'google-services.json'
}

```

4、iOS project 
 
 file: `ios/Flutter/Generated.xcconfig`

```
  facebookAppId=xxx
  versionCode=1
  android_applicationId=com.xxx.br
  ios_bundle_identifier=com.xxx.xxxx.nomodify
 ……
```

 file: `ios/Flutter/flutter_export_environment.sh`

```shell
export "facebookAppId=xxx"
export "versionCode=1"
export "android_applicationId=com.xxx.br"
export "ios_bundle_identifier=com.xxx.xxxx.nomodify"
……

```

You can use in Info.plist

```
	<key>CFBundleIdentifier</key>
	<string>$(ios_bundle_identifier)</string>
```

**Comment by maheshmnj** [comment]
Hi @hai046, Thanks for filing the issue. Labeling it a valid feature request.

**Comment by christopherfujino** [comment]
@hai046 thanks for filing this. I think this is a good idea. Some thoughts:

1. how about `--dart-define-from-file` for the option name?
2. out of curiosity, what is the value in exposing the raw BASE64 blob to Flutter apps?
3. will this support web and desktop as well?

**Comment by hai046** [comment]
@christopherfujino 
Great!
1、use `--dart-define-from-file`  naming is clear   👍🏻
2、Only raw_json, that is, the text content is base64, otherwise it will fail due to encoding problems when compiling
3、The web is supported during implementation. I have not fully tested the desktop version. I will study the integrity later.

**Comment by christopherfujino** [comment]
> 2、Only full file raw values that is, the text content is base64, otherwise it will fail due to encoding problems when compiling，other fields can be used directly

What I mean, is that what value is there to exposing the entire file to a Flutter app, when app code can always access individual fields which are already decoded. It seems like having the app decode base64 would be inefficient.

**Comment by hai046** [comment]
@christopherfujino 
This may just be a specification and scenario of our project: 
1. Centralized to a bean management, which is convenient for us to maintain; 
3. During the testing phase, testers can automatically switch some values, and only need to manage them in one place; 
4. 3. The pre-release and test environments can display all current configurations on one page, which is convenient for non-developers to preview and check

**Comment by christopherfujino** [comment]
@hai046 this sounds good for you to implement, as long as you ensure that this works for desktop and web builds.

**Comment by hai046** [comment]
@christopherfujino 
Thanks for your suggestion, it will be much easier to use if coding, I have optimized it

- You can use DEFINE_CONFIG_JSON_RAW_VALUE directly without base64
- FIx test cast
- New PR fix "Tree is currently broken. Please do not merge this PR unless it contains a fix for th"

**Comment by hai046** [comment]
I finished this last week, what do I need to do now, Or just wait for the review?

**Comment by christopherfujino** [comment]
> I finished this last week, what do I need to do now, Or just wait for the review?

Oops, sorry, I will review this today
